Ensure order is tracked by default

hmaarrfk opened the pull request:
This PR sets the order_tracked property to true by default, but for files that already exists, attempts to leave the property in the way it found them
- [x] Closes Issue https://github.com/h5netcdf/h5netcdf/issues/128
- [x] Tests added
    - [x] Basic file appending
    - [x] Editing a variable that already existed
    - [x] Adding a new variable in netcdf4 and reading it n h5netcdf (and vice versa)
- [x] Changes are documented in `CHANGELOG.rst`

kmuehlbauer commented on h5netcdf/core.py:
```diff
@@ -474,7 +485,7 @@ def _require_child_group(self, name):
     def create_group(self, name):
         if name.startswith("/"):
-            return self._root.create_group(name[1:])
+            return self._root.create_group(name[1:], track_order=self._track_order)
```
Here track_order isn't necessary, it's calling the (same) method of the root-object.

hmaarrfk replied:
Thanks!

kmuehlbauer commented on h5netcdf/core.py:
```diff
@@ -764,8 +777,49 @@ def _resize_variables(self, dimension, size, recurse=True):
     def __init__(
-        self, path, mode=None, invalid_netcdf=False, phony_dims=None, **kwargs
+        self, path, mode=None, invalid_netcdf=False, phony_dims=None,
+        *, track_order=True, **kwargs
```
Do we need the `*` after `phony_dims`?

hmaarrfk replied:
Not necessary, but gives you more flexibility in changing the API later since you can reorder keyword arguments, while you can't re-order the position of positional argiuments so easily.

It also guides people to write more readable code.

kmuehlbauer replied:
Great.

hmaarrfk replied:
This has been reverted now that the keyword argument has been removed.

kmuehlbauer commented on h5netcdf/core.py:
```diff
@@ -764,8 +777,49 @@ def _resize_variables(self, dimension, size, recurse=True):
+            Additional keyword arguments to be passed to the ``h5py.File``
+            constructor.
+        """
+        # 2022/01/09
```
We might just add these information to the issue and reference the issue.

hmaarrfk replied:
To be honest, these two sentences were to act like a short summary.
https://github.com/h5netcdf/h5netcdf/issues/128#issuecomment-1008396755

kmuehlbauer replied:
Let's leave it as is for now. Maybe good enough, if people just see it in the code.

kmuehlbauer commented on h5netcdf/tests/test_h5netcdf.py:
```diff
@@ -1387,3 +1401,32 @@ def test_no_circular_references(tmp_local_netcdf):
+
+
+def test_creation_with_h5netcdf_edit_with_netcdf4(tmp_local_netcdf):
+    # In version 0.12.0, the wrong file creation attributes were used
```
A test of the current known-fail would be good. So we know, when netcdf-c/netcdf4-python has implemented the new feature of being track_order agnostic.

hmaarrfk replied:
Added one.

hmaarrfk replied:
This is now rather hard to do because of the new error behavior. I've created an issue in this repo instead
https://github.com/h5netcdf/h5netcdf/issues/130

kmuehlbauer replied:
We could just utilize `h5py` instead of `h5netcdf` to create that feature.

kmuehlbauer replied:
The tracking issue is fine anyway.

hmaarrfk replied:
> We could just utilize h5py instead of h5netcdf to create that feature.

I'm not sure what this means.

kmuehlbauer replied:
Something along these lines:

```python
import h5py
import netCDF4
with h5py.File('test.nc', 'w', track_order=True) as f:
    pass
with netCDF4.Dataset("test.nc", mode="a") as f:
    pass
```

hmaarrfk replied:
Very smart. added. thanks!

kmuehlbauer reviewed (commented):
@hmaarrfk Just some comments to begin with.

hmaarrfk commented on CHANGELOG.rst:
```diff
@@ -144,4 +149,4 @@ Version 0.3.0 (Aug 7, 2016):
 - h5netcdf now writes the ``_NCProperties`` attribute for identifying netCDF4
-  files.
\ No newline at end of file
+  files.
```
I guess i added a new line at the end of the file. oops

kmuehlbauer commented:
@hmaarrfk Unfortunately the black style check doesn't show the diff. Apart from that LGTM.

hmaarrfk commented:
Great! Excited to see this in!

hmaarrfk commented on CHANGELOG.rst:
```diff
@@ -7,6 +7,11 @@ Development Version:
+- Ensure order tracking is true for newly created netcdf4 files as required
+  by the netcdf4 standard. A new keyword argument has been added to the class
+  constructor ``h5netcdf.File``, ``track_order`` to control the feature
+  for newly created files. (Closes Issue #128)
```
```suggestion
  for newly created files. This enables files created by h5netcdf to be
  appended to by netCDF4 library users. (Closes Issue #128)
```
Do we still have time to add this clarification?

kmuehlbauer replied:
Sure!

kmuehlbauer commented:
@hmaarrfk Thanks! I'll leave this open a couple of days to give @shoyer the chance to chime in.

hmaarrfk commented:
Sorry for rebasing. I added the suggested error behavior and updated the tests to match it.

I also added a test to ensure that the error behavior is as expected.

kmuehlbauer reviewed (approved):
@hmaarrfk Thanks again. I think this is ready to go. I'll merge tomorrow.

hmaarrfk commented:
I know you tagged this for a 1.0 release, though I know those are always contentious. Do you think it is possible to have a release before then?

kmuehlbauer commented:
@hmaarrfk I think it does no harm to release 0.13 after merging this.

Although there is a lot of effort to push 1.0 forward, it might take another while.

hmaarrfk commented:
> 0.13 after merging this.

Would you like me to update the text to indicate that this will come in version `0.13.0+` explicitely?

> 1.0 forward, it might take another while.

I totally understand mental blocks that exist before major versions.

kmuehlbauer commented:
> Would you like me to update the text to indicate that this will come in version `0.13.0+` explicitely?

That would be much appreciated, thanks.

hmaarrfk commented on h5netcdf/core.py:
```diff
@@ -766,6 +791,59 @@ class File(Group):
+        -----
+        In h5netcdf version 0.12.0 and earlier, order tracking was disabled in
+        HDF5 file. As this is a requirement for the current netCDF4 standard,
+        it has been enabled without deprecation in later versions [1]_.
```
```suggestion
        it has been enabled without deprecation as of version 0.13.0 [1]_.
```

hmaarrfk commented on h5netcdf/core.py:
```diff
@@ -766,6 +791,59 @@ class File(Group):
+                f"track_order, if specified must be set to to True (got {track_order})"
+                "to conform to the netCDF4 file format. Please see "
+                "https://github.com/h5netcdf/h5netcdf/issues/130 "
+                "for more details"
```
```suggestion
                "for more details."
```

kmuehlbauer commented:
Thanks again @hmaarrfk. I see that this is your first time contribution to `h5netcdf`. You are very welcome.
